pallet/dex/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+		#[pallet::call_index(1)]
+		pub fn withdraw(
+			origin: OriginFor<T>,
+			asset_id: u32,
+			amount: u128,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				TokenIndex::<T>::contains_key(asset_id),
+				Error::<T>::AssetIdNotInTokenIndex
+			);
+
+			ensure!(
+				UserTokenInfoes::<T>::contains_key(who.clone(), asset_id),
+				Error::<T>::AssetIdNotInTokenInfoes
+			);
+
+			let mut info = UserTokenInfoes::<T>::get(who.clone(), asset_id);
+			info.amount = info
+				.amount
+				.checked_sub(amount)
+				.ok_or(Error::<T>::TokenBalanceOverflow)?;
+
+			let mut sell_amout = 0;
+			for (who, order_index, _) in UserOrders::<T>::iter() {
+				let order = Orders::<T>::get(order_index).unwrap();
+
+				if order.address == who
+					&& order.pair.0 == asset_id
+					&& order.order_type == OrderType::SELL
+				{
+					sell_amout += order.amount_offered;
+				}
+			}
+
+			ensure!(
+				info.amount >= sell_amout,
+				Error::<T>::WithdrawBalanceMustKeepOrderSellAmount
+			);
+
+			<T::Fungibles as Mutate<T::AccountId>>::transfer(
+				asset_id,
+				&Self::account_id(),
+				&who,
+				amount,
+				Preservation::Expendable,
+			)?;
+
+			UserTokenInfoes::<T>::insert(who, asset_id, info);
+
+			Self::deposit_event(Event::Withdrawed { asset_id, amount });
+			Ok(().into())
+		}
+
+		#[pallet::weight({2})]
+		#[pallet::call_index(2)]
+		pub fn make_order(
+			origin: OriginFor<T>,
+			asset_id_1: u32,
+			asset_id_2: u32,
+			offered_amount: u128,
+			requested_amount: u128,
+			order_type: OrderType,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				asset_id_1 != asset_id_2,
+				Error::<T>::PairAssetIdMustNotEqual
+			);
+
+			NextOrderIndex::<T>::try_mutate(|index| -> DispatchResult {
+				let order_index = *index;
+
+				let order = Order {
+					counter: order_index,
+					pair: (asset_id_1, asset_id_2),
+					timestamp: 0,
+					order_type,
+					address: who.clone(),
+					amount_offered: offered_amount,
+					amout_requested: requested_amount,
+				};
+
+				*index = index
+					.checked_add(One::one())
+					.ok_or(Error::<T>::OrderIndexOverflow)?;
+
+				Orders::<T>::insert(order_index, &order);
+				UserOrders::<T>::insert(who.clone(), order_index, ());
+
+				let mut bounded_pair_orders = PairOrders::<T>::get((asset_id_1, asset_id_2));
+				bounded_pair_orders
+					.try_push(order_index)
+					.expect("Max bounded_pair_orders");
+				PairOrders::<T>::insert((asset_id_1, asset_id_2), bounded_pair_orders);
+
+				Self::deposit_event(Event::OrderCreated { order_index, order });
+				Ok(())
+			})?;
+
+			Ok(().into())
+		}
+
+		#[pallet::weight({3})]
+		#[pallet::call_index(3)]
+		pub fn cancel_order(origin: OriginFor<T>, order_index: u64) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			Orders::<T>::try_mutate_exists(order_index, |order| -> DispatchResult {
+				let order = order.take().ok_or(Error::<T>::InvalidOrderIndex)?;
```

Comment:
isn't take remove order from the storage? If so, any  can kill it

Reply:
yes, take is remove from orders

Reply:
i use "ensure!(order.address == who, Error::<T>::NotOwner);" to check order is belong of owner